#Exercise 10.03: Analysing numerical variables from the  Ames Housing dataset
In this exercise, we will continue our dataset exploration by analysing the numerical variables of this dataset. To do so, we will implement our own describe functions.

The dataset we will be using in this exercise is the Ames Housing dataset and it can be found on our GitHub repository - https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter10/dataset/ames_iowa_housing.csv

1. Open on a new Colab notebook and import the pandas package

In [0]:
import pandas as pd

2. Assign the link to the AMES dataset to a variable called 'file_url':

In [0]:
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter10/dataset/ames_iowa_housing.csv'

3. Using the read_csv method from the package pandas, load the dataset into a new variable called 'df':

In [0]:
df = pd.read_csv(file_url)

4. Create a new dataframe called 'num_df' with only the columns that are of numerical types using the method select_dtypes from pandas packages and pass in the value 'number' to the parameter 'include: 

In [0]:
num_df = df.select_dtypes(include='number')

5. Using the attribute **columns** from pandas, extract the list of columns of this dataframe 'num_df' and assign it to a new variable called 'num_cols' and print its content:

Expected Output:

![Figure 26 - List of numerical columns](https://docs.google.com/uc?export=download&id=1Xe6dR0fT6zT6-WIddLPvOOu2QU8OjMfX)



In [5]:
num_cols = num_df.columns
num_cols

Index(['Id', 'MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual',
       'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF',
       'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea',
       'MiscVal', 'MoSold', 'YrSold', 'SalePrice'],
      dtype='object')

6. Create a function called 'describe_numeric' that takes a **pandas** dataframe and a column name as input parameters. Then inside the function, print out the name of the given column, its minimum value using **min()**, maximum value using **max()**, its average value using **mean()**, standard deviation using **std()**, and median using **median()**:

In [0]:
def describe_numeric(df, col_name):
  print(f"\nCOLUMN: {col_name}")
  print(f"Minimum: {df[col_name].min()}")
  print(f"Maximum: {df[col_name].max()}")
  print(f"Average: {df[col_name].mean()}")
  print(f"Standard Deviation: {df[col_name].std()}")
  print(f"Median: {df[col_name].median()}")

7. Test this function by providing the 'df' dataframe and the column 'SalePrice':

Expected Output:

![Figure 30 - Display of the created function for the column ‘SalePrice’](https://docs.google.com/uc?export=download&id=1V0AZA0zxOcrbS2HLYFPhE75kBzy7CwPx)


In [7]:
describe_numeric(df, 'SalePrice')


COLUMN: SalePrice
Minimum: 34900
Maximum: 755000
Average: 180921.19589041095
Standard Deviation: 79442.50288288663
Median: 163000.0


The sale price ranges from 34,900 to 755,000 with an average of 180,921. The median is slightly lower than the average which tells us there are some outliers with high sales prices.

8. Create a for loop that will call the created function for every element from the list 'num_cols':

Expected Output:

![Figure 31 - Display of the created function for the first columns contained in ‘num_cols’](https://docs.google.com/uc?export=download&id=1N9LXQhRwfC6HlILm1kJBMa3xwax_Sydp)

In [8]:
for col_name in num_cols:
  describe_numeric(df, col_name)


COLUMN: Id
Minimum: 1
Maximum: 1460
Average: 730.5
Standard Deviation: 421.6100093688479
Median: 730.5

COLUMN: MSSubClass
Minimum: 20
Maximum: 190
Average: 56.897260273972606
Standard Deviation: 42.30057099381035
Median: 50.0

COLUMN: LotFrontage
Minimum: 21.0
Maximum: 313.0
Average: 70.04995836802665
Standard Deviation: 24.284751774483183
Median: 69.0

COLUMN: LotArea
Minimum: 1300
Maximum: 215245
Average: 10516.828082191782
Standard Deviation: 9981.264932379147
Median: 9478.5

COLUMN: OverallQual
Minimum: 1
Maximum: 10
Average: 6.0993150684931505
Standard Deviation: 1.3829965467415923
Median: 6.0

COLUMN: OverallCond
Minimum: 1
Maximum: 9
Average: 5.575342465753424
Standard Deviation: 1.1127993367127367
Median: 5.0

COLUMN: YearBuilt
Minimum: 1872
Maximum: 2010
Average: 1971.267808219178
Standard Deviation: 30.202904042525265
Median: 1973.0

COLUMN: YearRemodAdd
Minimum: 1950
Maximum: 2010
Average: 1984.8657534246574
Standard Deviation: 20.645406807709396
Median: 1994.0

COLUMN: Ma

The column ‘Id’ ranges from 1 to 1460 which is the exact value as the number of rows in this dataset so this column is definitely a unique identifier of the property sold. It seems the values from the ‘MSSubClass’ are all rounded. This may indicates the information contained in this column has either been clustered into groups of ten or may be a categorical variable.

Congratulations on completing this exercise! We saw how to explore a newly received dataset with just a few lines of code. This helped us to understand its structure, the type of information contained in each variable and also identified some potential data quality issues such as missing values or incorrect values.